File: ios/GliaWidgetsModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(GliaWidgetsModule, NSObject)

RCT_EXTERN_METHOD(multiply:(float)a withB:(float)b
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(configure:(NSString *) appToken
                  apiToken:(NSString *)apiToken
                  siteId:(NSString *)siteId
                  queueId:(NSString *)queueId)

RCT_EXTERN_METHOD(startChat)

RCT_EXTERN_METHOD(startAudioCall)

RCT_EXTERN_METHOD(startVideoCall)

@end

File: ios/GliaWidgetsModule.swift
import Foundation
import SalemoveSDK
import GliaWidgets

@objc(GliaWidgetsModule)
class GliaWidgetsModule: NSObject {

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc
    func configure(
        appToken: NSString,
        apiToken: NSString,
        siteId: NSString,
        queueId: NSString
    ) {

        let visitorContext = VisitorContext(
            type: .page,
            url: "http://www.glia.com"
        )

        let configuration = Configuration(
            appToken: String(appToken),
            apiToken: String(apiToken),
            environment: .usa,
            site: String(siteId)
        )
        startEngagement = { kind in

            DispatchQueue.main.async {
                do {
                    try Glia.sharedInstance.start(
                        kind,
                        configuration: configuration,
                        queueID: String(queueId),
                        visitorContext: visitorContext
                    )
                } catch {
                    print("Failed to configure Glia: \(error.localizedDescription)")
                }
            }

        }
    }

    @objc
    func startChat() {
        startEngagement?(.chat)
    }

    @objc
    func startAudioCall() {
        startEngagement?(.audioCall)
    }

    @objc
    func startVideoCall() {
        startEngagement?(.videoCall)
    }

    // MARK: - Private

    private var startEngagement: ((GliaWidgets.EngagementKind) -> Void)?
}
